{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
    "import operator\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cpg_positions (refseq, fwd_pos=31, rvs_pos=117):\n",
    "    pos=list()\n",
    "    for i,nuc in enumerate(refseq):\n",
    "        if nuc == 'C':\n",
    "            if refseq[i+1] == 'G':\n",
    "                if i>fwd_pos and i<rvs_pos: #exclude primers\n",
    "                    pos.append(i)\n",
    "    return(pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_reads (file):\n",
    "    d = {}\n",
    "    total_reads = 0\n",
    "    with open(file) as f:\n",
    "        for line in f:\n",
    "            val, seq = line.strip().split(\"\\t\")\n",
    "            d[seq] = int(val)\n",
    "            \n",
    "    return(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_alleles (d, refseq, min_freq=0.001):\n",
    "    #instantiate new dictionary\n",
    "    alleles = defaultdict(int)\n",
    "    \n",
    "    # count total number of reads\n",
    "    reads_n = sum(d.values())\n",
    "    \n",
    "    # get CpG positions\n",
    "    pos=get_cpg_positions(refseq)\n",
    "    \n",
    "    # amplicon length for 1st filter\n",
    "    refseq_len=len(refseq)\n",
    "    \n",
    "    #minimum number of reads to meet the min freq cutoff for 2nd filter\n",
    "    min_reads=reads_n*min_freq\n",
    "\n",
    "    for seq,val in d.items():\n",
    "        #1st and 2nd filters: exclude all indels, and minimum freq\n",
    "        if (len(seq) == refseq_len) & (val > min_reads):\n",
    "            allele=\"\"\n",
    "            for i,nuc in enumerate(seq):\n",
    "                if i in pos:\n",
    "                    allele+=nuc\n",
    "            testseq=set(allele)\n",
    "            #3rd filter, for any errors in the positions of interest\n",
    "            if 'A' not in testseq and 'G' not in testseq: \n",
    "                alleles[allele]+=val\n",
    "\n",
    "    # Sort by number of reads supporting the allele\n",
    "    alleles_sort = sorted(alleles.items(), key=operator.itemgetter(1), reverse=True)\n",
    "\n",
    "    # Total number of reads after filtering\n",
    "    filtered_reads=sum(alleles.values())\n",
    "    \n",
    "    return(alleles_sort,filtered_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_alleles_by_meCpG (alleles_sort):\n",
    "    d_group = defaultdict(int)\n",
    "    for allele,count in alleles_sort:\n",
    "        d_group[allele.count(\"C\")] +=count\n",
    "    return(d_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_df (alleles_sort,filtered_reads,freq_min):\n",
    "    # get CpG positions\n",
    "    pos=get_cpg_positions(refseq)\n",
    "    df=pd.DataFrame()\n",
    "    df_counts=pd.DataFrame()\n",
    "    for allele, val in alleles_sort:\n",
    "        freq=(val/filtered_reads*100)\n",
    "        if freq >= freq_min: #Frequency cut-offs\n",
    "            nuc_list = list(allele)\n",
    "            nuc_list.insert(0,allele)\n",
    "            row=pd.DataFrame(nuc_list).T\n",
    "            row.columns=(\"seq\",*pos)\n",
    "            df= df.append(row,ignore_index=True)\n",
    "            allele_counts = [allele, freq, val]\n",
    "            row_counts=pd.DataFrame(allele_counts).T\n",
    "            row_counts.columns=(\"seq\",\"freq\", \"val\")\n",
    "            df_counts = df_counts.append(row_counts,ignore_index=True)\n",
    "\n",
    "    df_melt = df.melt(id_vars = \"seq\")\n",
    "    df_melt_merged = pd.merge(df_melt, df_counts, on=\"seq\")\n",
    "    \n",
    "    return(df_melt_merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_meth_fraction (alleles_s, include_unmeth_alleles=True):\n",
    "    methylated_counts = defaultdict(int)\n",
    "    methylated_fraction = defaultdict(int)\n",
    "\n",
    "    totalreads = 0\n",
    "    for allele,val in alleles_s:\n",
    "        if include_unmeth_alleles or any('C' == nuc for nuc in allele):\n",
    "            totalreads += val\n",
    "            for i,nuc in enumerate(allele):\n",
    "                if nuc == 'C':\n",
    "                    methylated_counts[i]+=val\n",
    "                        \n",
    "    if not methylated_counts:\n",
    "        pos=get_cpg_positions(refseq)\n",
    "        for i,p in enumerate(pos):\n",
    "             methylated_counts[i]=0\n",
    "                \n",
    "\n",
    "    for i,methcount in methylated_counts.items():\n",
    "        if totalreads == 0:\n",
    "            methylated_fraction[i] = 0\n",
    "        else:\n",
    "            methfrac = methcount/totalreads\n",
    "            methylated_fraction[i] = methfrac\n",
    "\n",
    "\n",
    "    df = pd.DataFrame.from_dict(methylated_fraction, orient = \"index\")\n",
    "    \n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_meth_fraction_min (alleles_sort,filtered_reads,freq_min):\n",
    "    methylated_counts = defaultdict(int)\n",
    "    methylated_fraction = defaultdict(int)\n",
    "    \n",
    "    pos=get_cpg_positions(refseq)\n",
    "\n",
    "    val_sum=0\n",
    "    for allele, val in alleles_sort:\n",
    "        freq=(val/filtered_reads*100)\n",
    "        if freq < freq_min: #Frequency cut-offs and second filter, for any errors in the positions of interest\n",
    "            val_sum+=val\n",
    "            for i,nuc in enumerate(allele):\n",
    "                if nuc == 'C':\n",
    "                    methylated_counts[pos[i]]+=val\n",
    "                else:\n",
    "                    methylated_counts[pos[i]]+=0\n",
    "\n",
    "                    \n",
    "    for i,methcount in methylated_counts.items():\n",
    "        methfrac = methcount/val_sum\n",
    "        methylated_fraction[i] = methfrac\n",
    "\n",
    "    if not methylated_counts:\n",
    "        for i,p in enumerate(pos):\n",
    "             methylated_fraction[i]=0\n",
    "                \n",
    "    \n",
    "    df = pd.DataFrame.from_dict(methylated_fraction, orient = \"index\")\n",
    "    df.reset_index(level=0, inplace=True)\n",
    "    df.columns = ['variable', 'shade']\n",
    "    df['value'] = 'below_min'\n",
    "    df['seq'] = 'below_min'\n",
    "    df['freq'] = (val_sum/filtered_reads*100)\n",
    "    df['val'] = val_sum\n",
    "\n",
    "        \n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_lollipop_colour (df, outpath, outname=\"All_samples_combined_colour.pdf\"):  \n",
    "    # Changing default font to Arial\n",
    "    plt.rcParams['font.sans-serif'] = \"Arial\"\n",
    "    plt.rcParams['font.family'] = \"sans-serif\"\n",
    "    \n",
    "    df_melt = df.melt(id_vars=\"pos\")\n",
    "    df_melt['variable']= df_melt[\"variable\"].str.split('_').str[0]\n",
    "    df_melt = df_melt.sort_index(ascending=False)\n",
    "    #df_melt = df_melt.sort_values(by=['variable'])\n",
    "\n",
    "    plt.set_cmap('coolwarm')\n",
    "    plt.figure()\n",
    "    fig, ax = plt.subplots(figsize=(5,4))\n",
    "    ax.hlines(df_melt['variable'],min(df_melt['pos']),max(df_melt['pos']), label='_nolegend_', zorder=1)\n",
    "    im = ax.scatter(df_melt['pos'],df_melt['variable'],label=\"meC\", c=df_melt['value'],edgecolors =\"black\", s=50, zorder=2)\n",
    "    im.set_clim(0,1)\n",
    "    cbar = fig.colorbar(im, ax=ax,ticks=[0.1, 0.5, 0.9])\n",
    "    cbar.ax.tick_params(labelsize=6.5, # label size\n",
    "                        length=1.5, # length of ticks\n",
    "                        pad = 0.4) # distance between ticks and labels\n",
    "\n",
    "    ax.axes.set_xticks(list(df_melt['pos'].unique()))\n",
    "    ax.tick_params(axis='x', which='major', labelsize=6.5, rotation=45)\n",
    "    ax.tick_params(axis='y', which='major', labelsize=6.5)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    fig.savefig(outpath + outname)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_lollipop (df,sname,outpath,freq_min):\n",
    "    \n",
    "    # Changing default font to Arial\n",
    "    plt.rcParams['font.sans-serif'] = \"Arial\"\n",
    "    plt.rcParams['font.family'] = \"sans-serif\"\n",
    "    \n",
    "    plt.set_cmap('coolwarm')\n",
    "\n",
    "    df_C = df[df['value']==\"C\"]\n",
    "    df_T = df[df['value']==\"T\"]\n",
    "    \n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, sharey=True, figsize=(10,5))\n",
    "\n",
    "    ax1.hlines(df['seq'],min(df['variable']),max(df['variable']), label='_nolegend_', zorder=1)\n",
    "    ax1.scatter(df_C['variable'],df_C['seq'],label=\"meC\", color=\"#B30326\",edgecolors =\"black\", s=50, zorder=2)\n",
    "    ax1.scatter(df_T['variable'],df_T['seq'],label=\"C\", color=\"#3A4CC0\",edgecolors =\"black\", s=50, zorder=3)\n",
    "    ax1.legend(loc='lower center', bbox_to_anchor=(0.5, -0.3), ncol=2)\n",
    "    ax1.axes.set_xticks(list(df['variable'].unique()))\n",
    "    ax1.tick_params(axis='x', which='major', labelsize=6.5, rotation=45)\n",
    "    ax1.axes.set_xlabel(\"RAD51C CpG site\")\n",
    "\n",
    "    ax2.barh(df['seq'], df['freq'], align='center', color='grey')\n",
    "    ax2.axes.set_yticks([])\n",
    "    ax2.axes.set_xlabel(\"Frequency (%)\")\n",
    "    ax2.set_xlim([0, 100])\n",
    "\n",
    "    plt.suptitle(sname)# + f\"\\nMethylation alleles detected at >{freq_min}% frequency\") \n",
    "\n",
    "    fig.tight_layout(rect=[0, 0.03, 1, 0.9])\n",
    "\n",
    "    fig.savefig(f\"{outpath}{sname}_{freq_min}perc_barplot.pdf\")\n",
    "    \n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_lollipop_combined (df,df_below_freq,sname,outpath,freq_min, colbar=True):\n",
    "\n",
    "    # Changing default font to Arial\n",
    "    plt.rcParams['font.sans-serif'] = \"Arial\"\n",
    "    plt.rcParams['font.family'] = \"sans-serif\"\n",
    "\n",
    "    plt.set_cmap('coolwarm')\n",
    "\n",
    "    df_C = df[df['value']==\"C\"]\n",
    "    df_T = df[df['value']==\"T\"]\n",
    "\n",
    "    hb=df['seq'].drop_duplicates().count()\n",
    "\n",
    "    fig, ((ax3, ax4),(ax1, ax2)) = plt.subplots(2, 2, sharex='col',sharey='row',\n",
    "                                                gridspec_kw=dict(width_ratios=[10,5], height_ratios=[1,hb],hspace=0))    \n",
    "\n",
    "    # Merged epialleles\n",
    "    ax3.hlines(df_below_freq['seq'],min(df_below_freq['variable']),max(df_below_freq['variable']), label='_nolegend_', zorder=1)\n",
    "    im = ax3.scatter(df_below_freq['variable'],df_below_freq['seq'],label='_nolegend_', c=df_below_freq['shade'],edgecolors =\"black\", s=50, zorder=2)\n",
    "    im.set_clim(0,1)\n",
    "    ax3.tick_params(axis='x', which='major', labelsize=8, rotation=45)\n",
    "    ax3.axes.set_xticks([])\n",
    "\n",
    "    # Merged epialleles - frequency\n",
    "    ax4.barh(df_below_freq['seq'], df_below_freq['freq'], align='center', color='grey')\n",
    "    ax4.axes.set_yticks([])\n",
    "    ax4.set_xlim(0,100)\n",
    "\n",
    "    # Individual epialleles\n",
    "    ax1.hlines(df['seq'],min(df['variable']),max(df['variable']), label='_nolegend_', zorder=1)\n",
    "    ax1.scatter(df_C['variable'],df_C['seq'],label=\"meC\", color=\"#B30326\",edgecolors =\"black\", s=50, zorder=2)\n",
    "    ax1.scatter(df_T['variable'],df_T['seq'],label=\"C\", color=\"#3A4CC0\",edgecolors =\"black\", s=50, zorder=3)\n",
    "    ax1.axes.set_xticks(list(df['variable'].unique()))\n",
    "    ax1.tick_params(axis='x', which='major', labelsize=6.5, rotation=45)\n",
    "    ax1.axes.set_xlabel(\"RAD51C CpG site\",size=8,labelpad=4)\n",
    "    ax1.axes.set_yticks([])\n",
    "\n",
    "    # Individual epialleles - frequency\n",
    "    ax2.barh(df['seq'], df['freq'], align='center', color='grey')\n",
    "    ax2.axes.set_yticks([])\n",
    "    ax2.tick_params(axis='x', labelsize=6.5)\n",
    "    ax2.axes.set_xlabel(\"Frequency (%)\",size=8,labelpad=10)\n",
    "    ax2.set_xlim(0,100)\n",
    "\n",
    "    #Text labels\n",
    "    ax3.text(-0.135,0.4, f\"<{freq_min}% frequency \\n(merged)\", size=6.5, ha=\"center\", \n",
    "            transform=ax3.transAxes)\n",
    "    ax1.text(-0.14,0.9, f\"â‰¥{freq_min}% frequency\", size=6.5, ha=\"center\", \n",
    "            transform=ax1.transAxes)\n",
    "\n",
    "    # Figure title\n",
    "    fig.suptitle(sname, size=8, weight='bold')\n",
    "\n",
    "    fig.tight_layout(rect=[0, 0.03, 1, 0.9])\n",
    "    \n",
    "    if colbar:\n",
    "        # After tight layout, otherwise issues warnings\n",
    "        # Controlling the placement of the colour bar\n",
    "        adj_par = len(df.seq.unique())\n",
    "        adj_height=(2.5+adj_par*2.5)\n",
    "        adj_bbox_coord=(1.0436+0.0432*adj_par)\n",
    "\n",
    "        axins = inset_axes(ax3,\n",
    "                          width=\"60%\",  \n",
    "                         height=f\"{adj_height}%\",\n",
    "                          #height=\"7.5%\",  \n",
    "                          loc='lower left',\n",
    "                          bbox_to_anchor=(0, adj_bbox_coord, 1, 1), \n",
    "                          #bbox_to_anchor=(0, 1.25963685, 1, 1), \n",
    "                          bbox_transform=ax3.transAxes,\n",
    "                          borderpad=0)\n",
    "        # Colour bar\n",
    "        cbar = fig.colorbar(im, \n",
    "                            #label='CpG methylation',\n",
    "                            ax=ax3,\n",
    "                            cax=axins, \n",
    "                            ticks=[0.1, 0.5, 0.9],\n",
    "                            orientation=\"horizontal\")\n",
    "\n",
    "        cbar.ax.tick_params(labelsize=6.5, # label size\n",
    "                            length=1.5, # length of ticks\n",
    "                            pad = 0.4) # distance between ticks and labels\n",
    "\n",
    "        cbar.set_label(label='CpG methylation',\n",
    "                       labelpad=-25,\n",
    "                       size=8)\n",
    "        \n",
    "        fig.savefig(f\"{outpath}{sname}_{freq_min}perc_barplot.pdf\")\n",
    "    else:\n",
    "        fig.savefig(f\"{outpath}{sname}_{freq_min}perc_barplot_nolegend.pdf\")\n",
    "        \n",
    "\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !define reference sequence\n",
    "refseq=\"GAAAATTTACAAGACTGCGCAAAGCTGCAAGGCCCGGAGCCCCGTGCGGCCAGGCCGCAGA\\\n",
    "GCCGGCCCCTTCCGCTTTACGTCTGACGTCACGCCGCACGCCCCAGCGAGGGCGTGCGGAGTTTGGCTGCTCCGGGGTTAG\" \n",
    "# !define paths\n",
    "path=\"/Users/olgak/Documents/Postdoc_Waddell/Misc/2020_WFH/Scott_-_Ovarian_cancer/data/\"\n",
    "file_path=path + \"processed/amplicon_meth/\"\n",
    "outpath=file_path + \"notebook_filt_no_cutoff/\"\n",
    "\n",
    "if not os.path.exists(outpath):\n",
    "    os.mkdir(outpath)\n",
    "    \n",
    "    \n",
    "# !define sample ids to sample labels\n",
    "labels_df = pd.read_csv(path + \"supporting/SampleID_labels_amplicon_meth.csv\", index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_group_meCpG_comb = []\n",
    "\n",
    "for i,file in enumerate(glob.glob(file_path + \"*_grp\")):\n",
    "    # Extract sample label\n",
    "    sid=os.path.basename(file).replace(\"_L001_merged_seqprep.fastq.fna_1_num_grp\",\"\")\n",
    "    sname=labels_df.loc[sid]['ShortLabel']\n",
    "    if pd.isnull(sname):\n",
    "        sname=labels_df.loc[sid]['SampleLabel']\n",
    "    \n",
    "    # Generate dictionary with all reads\n",
    "    d=get_all_reads(file=file)\n",
    "\n",
    "    \n",
    "    # Count only CpG sites in alleles\n",
    "    alleles_sort,filtered_reads=count_alleles(d,refseq, min_freq=0)\n",
    "        \n",
    "    df_alleles_sort = pd.DataFrame(alleles_sort, columns=['allele',sname])\n",
    "\n",
    "    df_alleles_sort = df_alleles_sort.set_index('allele')\n",
    "    if i == 0:\n",
    "        df_alleles_sort_all=df_alleles_sort\n",
    "    elif i > 0:\n",
    "        df_alleles_sort_all=df_alleles_sort_all.join(df_alleles_sort, how=\"outer\")\n",
    "\n",
    "    #Calculate methylation fraction per CpG site\n",
    "    df_sample=calculate_meth_fraction(alleles_sort)\n",
    "    df_sample_unmeth=calculate_meth_fraction(alleles_sort, include_unmeth_alleles=False)\n",
    "    df_sample.columns=[sname]\n",
    "    df_sample_unmeth.columns=[sname]\n",
    "    if i == 0:\n",
    "        df_alt=df_sample\n",
    "        df_alt_unmeth=df_sample_unmeth\n",
    "    elif i > 0:\n",
    "        df_alt=df_alt.join(df_sample)\n",
    "        df_alt_unmeth=df_alt_unmeth.join(df_sample_unmeth)\n",
    "\n",
    "    # Prepare individual sample plots grouping alleles <2% and 5%\n",
    "    for freq_min in [5]:\n",
    "        df=convert_to_df(alleles_sort,filtered_reads,freq_min)\n",
    "        df_below_freq=calculate_meth_fraction_min(alleles_sort,filtered_reads,freq_min)\n",
    "        df.variable = df.variable - 156\n",
    "        df_below_freq.variable = df_below_freq.variable - 156\n",
    "        plot_path=f'{outpath}{freq_min}_perc/'\n",
    "        if not os.path.exists(plot_path):\n",
    "            os.mkdir(plot_path)\n",
    "        if df_below_freq.freq.sum() > 0:           \n",
    "            plot_lollipop_combined(df,df_below_freq,sname,plot_path,freq_min)\n",
    "            plot_path2=f'{outpath}{freq_min}_perc_no_legend/'\n",
    "            if not os.path.exists(plot_path2):\n",
    "                os.mkdir(plot_path2) \n",
    "            plot_lollipop_combined(df,df_below_freq,sname,plot_path2,freq_min, colbar=False)\n",
    "        else:\n",
    "            plot_lollipop(df,sname,plot_path,freq_min)\n",
    "            \n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_alleles_sort_all.to_csv(outpath + \"df_alleles_sort_all.csv\")\n",
    "df_alt.to_csv(outpath + \"df_meth_freq.csv\")\n",
    "df_alt_unmeth.to_csv(outpath + \"df_exclude_unmeth-alleles_freq.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_alt = df_alt.filter(labels_df.dropna().ShortLabel.tolist())\n",
    "\n",
    "df_alt_unmeth = df_alt_unmeth.filter(labels_df.dropna().ShortLabel.tolist())\n",
    "\n",
    "pos_promoter = list(map(lambda x: x - 156, get_cpg_positions(refseq)))\n",
    "df_alt['pos'] = pos_promoter\n",
    "df_alt_unmeth['pos'] = pos_promoter\n",
    "\n",
    "\n",
    "plt.style.use('default')\n",
    "plot_lollipop_colour(df=df_alt, outpath=outpath,\n",
    "                    outname=\"All_samples_combined_colour.pdf\")\n",
    "\n",
    "plot_lollipop_colour(df=df_alt_unmeth, outpath=outpath,\n",
    "                     outname=\"All_samples_combined_exclude_unmeth-alleles_colour.pdf\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
